kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h

```diff
@@ -933,37 +955,45 @@ class ResidualBasedBlockBuilderAndSolver
     ///@}
     ///@name Protected Operations
     ///@{
+
     void ConstructMasterSlaveConstraintsStructure(ModelPart &rModelPart)
     {
         if (rModelPart.MasterSlaveConstraints().size() > 0) {
             const ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
 
+            // Vector containing the localization in the system of the different terms
+            DofsVectorType slave_dof_list, master_dof_list;
+
             // Constraint initial iterator
             const auto const_begin = rModelPart.MasterSlaveConstraints().begin();
-            std::vector<std::unordered_set<int>> indices(BaseType::mDofSet.size());
+            std::vector<std::unordered_set<IndexType>> indices(BaseType::mDofSet.size());
 
             std::vector<LockObject> lock_array(indices.size());
 
-            #pragma omp parallel
+            #pragma omp parallel firstprivate(slave_dof_list, master_dof_list)
             {
                 Element::EquationIdVectorType slave_ids(3);
                 Element::EquationIdVectorType master_ids(3);
-                std::unordered_map<int, std::unordered_set<int>> temp_indices;
+                std::unordered_map<IndexType, std::unordered_set<IndexType>> temp_indices;
 
                 #pragma omp for schedule(guided, 512) nowait
                 for (int i_const = 0; i_const < static_cast<int>(rModelPart.MasterSlaveConstraints().size()); ++i_const) {
                     auto it_const = const_begin + i_const;
                     it_const->EquationIdVector(slave_ids, master_ids, r_current_process_info);
+                    it_const->GetDofList(slave_dof_list, master_dof_list, r_current_process_info);
+
                     // Slave DoFs
-                    for (auto &id_i : slave_ids)
-                    {
-                        temp_indices[id_i].insert(master_ids.begin(), master_ids.end());
+                    for (auto &id_i : slave_ids) {
+                        for (auto& master_dof : master_dof_list) {
+                            if (master_dof->IsFixed() || BaseType::mDofSet.find(*master_dof) != BaseType::mDofSet.end()) {

```

Comment:
note that i am adding this anyway... i am adding more but that behaviour is intentional

Reply:
reformulating my comment: why should i avoid adding one of the master_dofs? 
In my version i am always adding it, i don't see how this could harm